File: 322.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
  int coinChange(vector<int> &coins, int amount) {
    if (amount == 0)
      return 0;
    vector<int> dp = vector<int>(amount + 1, amount + 1);
    dp[0] = 0;
    for (int i = 1; i <= amount; i++)
      for (int coin : coins)
        if (i - coin < 0)
          continue;
        else
          dp[i] = min(dp[i], dp[i - coin] + 1);
    return dp[amount] == amount + 1 ? -1 : dp[amount];
  }
};

File: python/322.py
from typing import List


class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        nums = [1e4+1 for _ in range(amount+1)]
        nums[0] = 0

        for i in range(1, amount+1):
            for c in coins:
                if c > amount:
                    continue
                nums[i] = min(nums[i], nums[i-c]+1)

        if nums[amount] == 10001:
            return -1
        return nums[amount]
